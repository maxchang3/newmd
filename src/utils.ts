import yaml from 'js-yaml'
import TOML from 'smol-toml'

export const getFrontmatter = (object: Record<string, unknown>, toml: boolean) => {
    const fence = toml ? '+++' : '---'
    const frontmatter = toml ? TOML.stringify(object) : yaml.dump(object).trim()
    return `${fence}\n${frontmatter}\n${fence}\n`
}
