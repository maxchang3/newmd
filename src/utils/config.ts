import type { Config } from '@/types'
import { DEFAULT_CONFIG } from '@/consts'
import { deepmerge } from 'deepmerge-ts'
import { createConfigLoader } from 'unconfig'

export const resolveConfig = async (config?: Config, options?: {
    cwd?: string
}) => {
    const loader = createConfigLoader<Config>({
        sources: [
            {
                files: 'newmd.config',
                extensions: ['ts', 'mjs', 'js'],
            },
        ],
        cwd: options?.cwd,
    })

    const loadResult = await loader.load()

    // Respect config file when no explicit options are provided
    if (!loadResult.sources.length) return deepmerge(DEFAULT_CONFIG, config) as Required<Config>

    return deepmerge(deepmerge(DEFAULT_CONFIG, loadResult.config), config) as Required<Config>
}
