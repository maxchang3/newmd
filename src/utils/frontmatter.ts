import type { StringKeys, ZodStringKeys } from '@/types'
import type { AnyZodObject } from 'zod'
import yaml from 'js-yaml'
import toml from 'smol-toml'
import { init as createSchemaDefaults } from 'zod-empty'

export type FrontmatterType = 'yaml' | 'toml'

interface FrontmatterConfig<Metadata extends Record<string, any>> {
    /**
     * Base metadata for the markdown file, will be used to generate the frontmatter.
     */
    baseMetadata: Metadata
    /**
     * Key in the metadata object that will be used as the title, the value must be a string.
     */
    titleKey: StringKeys<Metadata>
    /**
     * Type of frontmatter to generate.
     *
     * @defaultValue "yaml"
     */
    type?: FrontmatterType
}

export class Frontmatter<Metadata extends Record<string, any>> {
    config: FrontmatterConfig<Metadata>

    constructor({ baseMetadata, titleKey, type = 'yaml' }: typeof this.config) {
        this.config = { baseMetadata, type, titleKey }
    }

    static fromZodSchema<Schema extends AnyZodObject>(schema: Schema, options: {
        titleKey: ZodStringKeys<Schema>
        type?: FrontmatterType
    }) {
        const { titleKey, type = 'yaml' } = options
        const defaultData = createSchemaDefaults(schema)
        return new Frontmatter({ baseMetadata: defaultData, type, titleKey })
    }

    toString(title: string) {
        const { titleKey, baseMetadata } = this.config
        const metadata = structuredClone(baseMetadata)
        const filteredMetadata = Object.fromEntries(
            Object.entries(metadata).filter(([_, value]) => value !== null),
        )
        if (titleKey in filteredMetadata) {
            // @ts-expect-error titleKey is a valid key in metadata
            filteredMetadata[titleKey] = title
        } else {
            throw new Error(`Title key "${String(titleKey)}" not found in metadata`)
        }
        const isTOML = this.config.type === 'toml'
        const fence = isTOML ? '+++' : '---'
        const frontmatter = isTOML
            ? toml.stringify(filteredMetadata)
            : yaml.dump(filteredMetadata).trim()
        return `${fence}\n${frontmatter}\n${fence}\n`
    }
}
