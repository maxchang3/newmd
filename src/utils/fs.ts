import fs from 'node:fs/promises'
import { resolve } from 'pathe'

export const writeMarkdownFile = async (options: {
    filename: string
    frontmatter: string
    content?: string
    path?: string
    cwd?: string
}) => {
    const { filename, frontmatter, content, path, cwd } = options

    const outputDir = resolve(cwd ?? '', path ?? '.')
    const filepath = resolve(outputDir, `${filename}.md`)

    let fileContent = frontmatter
    if (content) fileContent += `\n${content}`

    await fs.mkdir(outputDir, { recursive: true })
    await fs.writeFile(filepath, fileContent)

    return filepath
}
