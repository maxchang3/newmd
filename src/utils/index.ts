export * from './config'
export * from './frontmatter'
export * from './fs'
