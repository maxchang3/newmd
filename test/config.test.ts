import { resolveConfig } from '@/utils'
import { resolve } from 'pathe'
import { describe, expect, it } from 'vitest'

const fixtureDir = resolve(__dirname, 'fixtures')

describe('config', () => {
    it('should return default config', async () => {
        const config = await resolveConfig({}, { cwd: fixtureDir })
        expect(config).matchSnapshot()
    })

    it('should return merged config', async () => {
        const cwd = resolve(fixtureDir, 'headline')
        const config = await resolveConfig({}, { cwd })
        expect(config).matchSnapshot()
    })

    it('should support schema-specific paths', async () => {
        const cwd = resolve(fixtureDir, 'schema-paths')
        const config = await resolveConfig({}, { cwd })

        // Test that path config is resolved as object form with all schemas
        expect(config.path).toMatchSnapshot('schema-specific-paths')
        expect(config.format).toBe('yaml')
        expect(config.titleMapping).toBe('title')
    })
})
