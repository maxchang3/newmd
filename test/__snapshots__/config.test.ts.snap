// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`config > should return default config 1`] = `
{
  "path": "./src/data/blog",
  "schemas": {
    "blog": ZodObject {
      "_cached": null,
      "_def": {
        "catchall": ZodNever {
          "_def": {
            "typeName": "ZodNever",
          },
          "and": [Function],
          "array": [Function],
          "brand": [Function],
          "catch": [Function],
          "default": [Function],
          "describe": [Function],
          "isNullable": [Function],
          "isOptional": [Function],
          "nullable": [Function],
          "nullish": [Function],
          "optional": [Function],
          "or": [Function],
          "parse": [Function],
          "parseAsync": [Function],
          "pipe": [Function],
          "promise": [Function],
          "readonly": [Function],
          "refine": [Function],
          "refinement": [Function],
          "safeParse": [Function],
          "safeParseAsync": [Function],
          "spa": [Function],
          "superRefine": [Function],
          "transform": [Function],
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        "shape": [Function],
        "typeName": "ZodObject",
        "unknownKeys": "strip",
      },
      "and": [Function],
      "array": [Function],
      "augment": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nonstrict": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
  },
  "titleMapping": "title",
  "toml": false,
}
`;

exports[`config > should return merged config 1`] = `
{
  "path": "./src/data/blog",
  "schemas": {
    "blog": ZodObject {
      "_cached": null,
      "_def": {
        "catchall": ZodNever {
          "_def": {
            "typeName": "ZodNever",
          },
          "and": [Function],
          "array": [Function],
          "brand": [Function],
          "catch": [Function],
          "default": [Function],
          "describe": [Function],
          "isNullable": [Function],
          "isOptional": [Function],
          "nullable": [Function],
          "nullish": [Function],
          "optional": [Function],
          "or": [Function],
          "parse": [Function],
          "parseAsync": [Function],
          "pipe": [Function],
          "promise": [Function],
          "readonly": [Function],
          "refine": [Function],
          "refinement": [Function],
          "safeParse": [Function],
          "safeParseAsync": [Function],
          "spa": [Function],
          "superRefine": [Function],
          "transform": [Function],
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        "shape": [Function],
        "typeName": "ZodObject",
        "unknownKeys": "strip",
      },
      "and": [Function],
      "array": [Function],
      "augment": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nonstrict": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
  },
  "titleMapping": "headline",
  "toml": false,
}
`;
